### DISCUSSION TITLE
Tutorial - 5. Testing contracts

### DISCUSSION THREAD

**Discussion by ridhof** [NomicFoundation/hardhat#3968] [Troubleshooting]
Hi, I was following the tutorial here, [https://hardhat.org/tutorial/testing-contracts](https://hardhat.org/tutorial/testing-contracts). 
I was a bit confused when its informed that the test are going to use ethers.js and Mocha, but there is no npm install instruction like the previous page.

and then I installed ethers with `npm i --save-dev ethers`.
and then I tried to install mocha with `npm i --save-dev mocha` but then I got this error log

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @nomicfoundation/hardhat-chai-matchers@1.0.6
npm ERR! Found: ethers@6.4.0
npm ERR! node_modules/ethers
npm ERR!   dev ethers@"^6.4.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer ethers@"^5.0.0" from @nomicfoundation/hardhat-chai-matcher
npm ERR! node_modules/@nomicfoundation/hardhat-chai-matchers
npm ERR!   peer @nomicfoundation/hardhat-chai-matchers@"^1.0.0" from @no
npm ERR!   node_modules/@nomicfoundation/hardhat-toolbox
npm ERR!     dev @nomicfoundation/hardhat-toolbox@"^2.0.2" from the root
npm ERR! 
npm ERR! Conflicting peer dependency: ethers@5.7.2
npm ERR! node_modules/ethers
npm ERR!   peer ethers@"^5.0.0" from @nomicfoundation/hardhat-chai-match
npm ERR!   node_modules/@nomicfoundation/hardhat-chai-matchers
npm ERR!     peer @nomicfoundation/hardhat-chai-matchers@"^1.0.0" from @
npm ERR!     node_modules/@nomicfoundation/hardhat-toolbox
npm ERR!       dev @nomicfoundation/hardhat-toolbox@"^2.0.2" from the ro
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency reso
```

and then I checked my `package-lock.json` and I found mocha is already installed there. 
Installing ethers is also include mocha installation, I assume.

and then, I run the `npx hardhat test`, and I got this error:
```


  Token contract
    1) Deployment should assign the total supply of tokens to the owner


  0 passing (418ms)
  1 failing

  1) Token contract
       Deployment should assign the total supply of tokens to the owner:
     TypeError: Cannot read properties of undefined (reading 'JsonRpcProvider')
      at Object.<anonymous> (node_modules/@nomiclabs/hardhat-ethers/src/internal/ethers-provider-wr
pper.ts:4:61)
      at Module._compile (node:internal/modules/cjs/loader:1105:14)
      at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
      at Module.load (node:internal/modules/cjs/loader:981:32)
      at Function.Module._load (node:internal/modules/cjs/loader:822:12)
      at Module.require (node:internal/modules/cjs/loader:1005:19)
      at require (node:internal/modules/cjs/helpers:102:18)
      at Object.<anonymous> (node_modules/@nomiclabs/hardhat-ethers/src/internal/provider-proxy.ts:
:1)
      at Module._compile (node:internal/modules/cjs/loader:1105:14)
      at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
```

I am assuming that the `ethers` are not imported properly or something, but I found the same issue posted in here [https://ethereum.stackexchange.com/questions/144451/typeerror-cannot-read-properties-of-undefined-reading-jsonrpcprovider](https://ethereum.stackexchange.com/questions/144451/typeerror-cannot-read-properties-of-undefined-reading-jsonrpcprovider).

Following the solution, I tried to downgrade, and then its successfully run the test.

I think it will be great if the tutorial could also include the installation sample to be consistent like the previous chapter.
and also I think it will be great if including the ether.js version that is working.

**Comment by fvictorio** [comment]
Hi @ridhof, the reason why ethers is not installed explicitly is that you install the Hardhat toolbox, which has ethers as a peer dependency. npm, starting from version 7, installs peer dependencies automatically.

If you are using yarn or an older version of npm, then you do have to install everything manually. Hardhat docs have tabs to select the package manager you are using:

![image](https://github.com/NomicFoundation/hardhat/assets/417134/01fd446a-12a2-4490-9038-32a5ee9b4206)

If you select `npm 6` or `yarn`, then you'll have the full list of packages (and in that snippet, ethers is indicated as `ethers@5`, to prevent you from installing the latest major version, which doesn't fully work with Hardhat yet).

**Reply by fvictorio** [reply]
Ah, but there is an error in those instructions: the `hardhat-ethers` plugin should also use the previous major version. I'll fix that!

**Reply by fvictorio** [reply]
No, wait, that's also fine because the new version of the plugin has a different org. So that command should work, but please let me know if it doesn't.

**Comment by shamras** [comment]
how do I downgrade my chain in npm its currently sitting in a mocha test project but how do I change the current chai version to something older?

**Reply by schaable** [reply]
`npm i chai@version`, for example `npm i chai@4.0.0`
